theano/contrib/mkl/mkl_type.py

```diff
@@ -0,0 +1,350 @@
+    or not (to be broadcastable a dimension must always be of length 1)
+
+    The `context_name` is the name of the context. For similar interface
+    with the new back-end for GPU.
+
+    `name` for the type that will be used in printouts.
+
+    """
+    context_name = 'mkl'
+    Variable = MKLNdarrayVariable
+    Constant = None
+    SharedVariable = None
+    ndim = None
+    broadcastable = None
+
+    def __init__(self, dtype, broadcastable, context_name=None, name=None):
+        self.dtype = str(dtype)
+        if self.dtype == 'floatX':
+            self.dtype = config.floatX
+
+        if self.dtype not in ('float32', 'float64'):
+            raise TypeError('%s only supports float32/float64 for now.'
+                            'Tried using dtype %s for variable %s' %
+                            (self.__class__.__name__, dtype, name))
+
+        self.typenum = 11 if dtype is 'float32' else 12
+        self.broadcastable = tuple(bool(b) for b in broadcastable)
+        self.name = name
+        self.dtype_specs()
+        if not (context_name is None):
+            self.context_name = context_name
+
+    def clone(self, dtype=None, broadcastable=None):
+        if broadcastable is None:
+            broadcastable = self.broadcastable
+
+        if dtype is None:
+            dtype = self.dtype
+
+        return self.__class__(dtype=dtype, broadcastable=broadcastable,
+                              context_name=self.context_name, name=self.name)
+
+    def filter(self, data, strict=False, allow_downcast=None):
+        """
+        Since it is not support currently to cast layout from FP32 to FP64, we
+        don't do up or down cast in this function for now. It means that it is
+        always strict for all input data.
+
+        """
+        # TODO: we will add up and down cast code here when the low level APIs
+        # support that.
+        if not isinstance(data, mkl.MKLNdarray):
+            raise TypeError('%s expected a MKLNdarray object.' % self,
+                            data, type(data))
+
+        if data.dtype != self.dtype:
+            raise TypeError(('%s expected a MKLNdarray object with'
+                            'dtype=%s (got %s).') % (self, self.dtype, data.dtype))
+
+        if data.ndim != self.ndim:
+            raise TypeError('Wrong number of dimensions: expected %s,'
+                            'got %s with shape %s.' % (self.ndim, data.ndim, data.shape))
+
+        shp = data.shape
+        for i, b in enumerate(self.broadcastable):
+            if b and shp[i] != 1:
+                raise TypeError('None-unit value on shape on a broadcastable'
+                                ' dimension.', shp, self.broadcastable)
+
+        return data
+
+    def filter_variable(self, other, allow_convert=True):
+        raise NotImplementedError('MKLNdarrayType filter_variable')
+
+    @staticmethod
+    def values_eq(a, b):
+        raise NotImplementedError('MKLNdarray values_eq')
```

Comment:
This probably need to be implemented. values_eq_approx() must be implemented to be able to work in DebugMode. At worst, convert both object to ndarray and call TensorType.values_eq_approx() on the converted version.